Improvement to c_enum_typet interface

esteffin opened the pull request:
This PR improves the `c_enum_typet` interface to avoid having to directly access its `sub` components to add members.
To achieve this it:
 - adds a new `members()` function returning a non-`const` reference
 - adds a new constructor that has 2 parameters: the underlying type and a `std::vector` of members to populate the enumeration with.

The PR also refactor the code where the `c_enum_typet` values are populated in `c_typecheck_type` to use the new `members` function.

Commit:
Use new improved functions for c_enum_typet

NlightNFotis commented on src/util/c_types.h:
```diff
@@ -274,6 +274,17 @@ class c_enum_typet : public type_with_subtypet
+
+  memberst &members()
+  {
+    return (memberst &)(add(ID_body).get_sub());
```
Why does this `add` an `ID_body` whereas the `const` version of this does a `find` for the same thing?

Is it possible that this (where *this* == `add`) fails in some capacity? Say if an `ID_body` already exists? Or am I misunderstanding this?

thomasspriggs replied:
`irept::add` will add the given key to the named subs if it doesn't exist and return the corresponding value in either case. The corresponding call to `irept::find` in the `const` overload will return a fresh nil irept in the case where `ID_body` is not in the map. `irept::add` cannot be called from a `const` context because it is not `const` itself, because it adds the key.

esteffin replied:
The `find` function returns a `const` reference.
`add` instead creates the element if it does not exists (and it does not fail, re-create or delete it if it exists) and returns it as a non-`const` reference to the (possibly newly created) object at the given key (`ID_body` in this case).

NlightNFotis commented on src/util/c_types.h:
```diff
@@ -274,6 +274,17 @@ class c_enum_typet : public type_with_subtypet
+
   const memberst &members() const
   {
     return (const memberst &)(find(ID_body).get_sub());
```
I know you didn't write those, but wouldn't a C++ style cast (either a `static_cast` or `reinterpret_cast` be a better idea in both of the `members()` methods?

thomasspriggs replied:
This would nominally be a `static_cast`, for the usual casting pattern we use with `irept`s.

thomasspriggs replied:
I was mistaken. The correct cast to use here is `reinterpret_cast`. `static_cast` work work for casting around the `irept` heirarchy directly, however the types here are two different specialisations of `std::vector`, rather than two types in the same hierarchy. Hence the `reinterpret_cast` is needed.

NlightNFotis commented on src/ansi-c/c_typecheck_type.cpp:
```diff
@@ -1413,14 +1414,11 @@ void c_typecheck_baset::typecheck_c_enum_type(typet &type)
   enum_tag_symbol.type.add_subtype() = underlying_type;
 
+  // throw in the enum members as 'body'
+  to_c_enum_type(enum_tag_symbol.type).members() = std::move(enum_members);
```
👍🏻

Commit:
Improved interface for c_enum_typet

Interface for c_enum_typet has limited support for management of enum
members.
This commit adds:
- a function for exposing the member vector as a non-const reference,
- a constructor that takes the underlying type and a list of members and
  sets the constructed type accordingly.

thomasspriggs commented on unit/solvers/smt2_incremental/construct_value_expr_from_smt.cpp:
```diff
@@ -36,7 +36,8 @@ static type_symbolt make_c_enum_type_symbol(std::size_t underlying_size)
 
-  auto &members = enum_type.add(ID_body).get_sub();
+  auto &members = enum_type.members();
+  members.reserve(20);
```
⛏️ I might have stored `20` in a constant instead of duplicating the same magic number in 2 places.
⛏️ Also `reserve` is used for performance reasons. So maybe it is better to omit it in a unit test context in order to aid readability.
